src/sagemaker/mp_config.py

```diff
@@ -0,0 +1,270 @@
+"""Parses and validates configurations for modelparallel arguments.
+
+This validation is done in Python SDK so that validation errors are
+immediately raised, before launching the training job.
+"""
+from __future__ import absolute_import
+
+# Standard Library
+import json
+import os
+import re
+from pydoc import locate
+
+# Third Party
+import yaml
+
+# NOTE: this file is duplicated across smp and SageMaker Python SDK.
+# any changes to this file requires an update of the corresponding file
+# in SageMaker Python SDK upon release.
+
+try:
+    from smdistributed.modelparallel.backend.logger import get_logger
+
+    logger = get_logger()
+    info_func = logger.info
+    warn_func = logger.warning
+    py_sdk = False
+except ImportError:
+    # the case where this file runs on Python SDK for config validation
+    info_func = print
+    warn_func = print
+    py_sdk = True
+
+
+def _int_or_float_if_possible(s):
+    """Attempt to convert to int, or float"""
+    try:
+        float_s = float(s)
+        int_s = int(float_s)
+        return int_s if int_s == float_s else float_s
+    except ValueError:
+        return s
+
+
+class ConfigParam:
+    """Represents and validates a single modelparallel configuration parameter."""
+
+    def __init__(self, name, input_value, cfg_dict, existing_params, provided=False):
+        self.name = name
+        self.cfg_dict = cfg_dict
+        self.existing_params = existing_params
+        self.provided = provided
+        self._value = self._set_value(input_value)
+
+    def get_value(self):
+        """Return the underlying value of the configuration parameter"""
+        return self._value
+
+    def _get_default(self):
+        """Get default value for the parameter"""
+        default = self._handle_dependencies(self.cfg_dict["default"])
+        if isinstance(default, (float, int)):
+            # should explicitly enforce upper and lower bounds for dynamic formulas in case
+            # the default evaluates out of bounds
+            if "upper_bound" in self.cfg_dict:
+                default = min(default, self._handle_dependencies(self.cfg_dict["upper_bound"]))
+            if "lower_bound" in self.cfg_dict:
+                default = max(default, self._handle_dependencies(self.cfg_dict["lower_bound"]))
+
+        return default
+
+    def _set_value(self, input_value):
```

Comment:
Can we use python properties here instead of getter/setter ?
Also this func is too long, can we make this more modular?